nv_tiles.py

```diff
@@ -0,0 +1,200 @@
+# Based on https://github.com/envytools/envytools/blob/master/nvhw/tile.c
+#FIXME: Add https://github.com/envytools/envytools/blob/master/hwtest/nv10_tile.cc#L35
+
+def is_igp(chipset):
+  #  switch (chipset) {
+  #    case 0x0a:
+  #    case 0x1a:
+  #    case 0x1f:
+  #    case 0x2a:
+  #    case 0x4e:
+  #    case 0x4c:
+  #    case 0x67:
+  #    case 0x68:
+  #    case 0x63:
+  #    case 0xaa:
+  #    case 0xac:
+  #    case 0xaf:
+  #      return 1;
+  #    default:
+  #      return 0;
+  #  }
+  return True
+
+def has_tile_factor_13(chipset):
+  # return pfb_type(chipset) >= PFB_NV20 && chipset != 0x20
+  return True
+
+def has_large_tile(chipset):
+  # return pfb_type(chipset) == PFB_NV40 || pfb_type(chipset) == PFB_NV41
+  return False
+
+PFB_NV10 = 1
+PFB_NV20 = 2
+
+def pfb_type(chipset):
+  return PFB_NV20
+
+def tile_pitch_valid(chipset, pitch):
+
+  if (pitch & ~0x1ff00):
+    return False, 0, 0
+
+  pitch >>= 8
+  if (pitch == 0):
+    return False, 0, 0
+
+  if (pitch == 1):
+    return False, 0, 0
+  if (pitch & 1) and ((pfb_type(chipset) == PFB_NV10) or (pfb_type(chipset) == PFB_NV44)):
+    return False, 0, 0
+  if (pitch > 0x100):
+    return False, 0, 0
+
+  shift = 0
+  while ((pitch & 1) == 0):
+    pitch >>= 1
+    shift += 1
+
+  if (shift >= 8) and has_large_tile(chipset) == False:
+    return False, 0, 0
+
+  if (pitch == 1):
+    factor = 0
+  elif (pitch == 3):
+    factor = 1
+  elif (pitch == 5):
+    factor = 2
+  elif (pitch == 7):
+    factor = 3
+  elif (pitch == 13):
+    if has_tile_factor_13(chipset) == False:
+      return False, 0, 0
+    factor = 4
+  else:
+    return False, 0, 0
+
+  return True, shift, factor
+
+class mc_config:
+  def __init__(self):
+    self.mcbits = 0
+    self.partbits = 0
+    self.colbits_lo = 0
+    self.burstbits = 0
+
+def tile_translate_addr(chipset, pitch, address, mode, bankoff, mcc):
+  bankshift = mcc.mcbits + mcc.partbits + mcc.colbits_lo
+  is_vram = (mode == 1) or (mode == 4)
+  if (is_igp(chipset)):
+    is_vram = 0
+  if (is_vram == False):
+    bankshift = 12
+
+  valid, shift, factor = tile_pitch_valid(chipset, pitch)
+  if (valid == False):
+    assert(False)
+
+  x = address % pitch
+  y = address // pitch
+  ix = x & 0xff
+  iy = y & ((1 << (bankshift - 8)) - 1)
+
+  x >>= 8
+  y >>= bankshift - 8
+  iaddr = 0
+  baddr = y * (pitch >> 8) + x
+  part = 0
+  tag = 0
+
+  fb_type = pfb_type(chipset)
+  if fb_type == PFB_NV10:
```

Comment:
This codepath can probably be removed. If this is bad, blame it all on mooch

Reply:
lol fair enough